src/Streamly/Mem/Array.hs

```diff
@@ -218,28 +222,81 @@ _last arr = readIndex arr (length arr - 1)
 -- Searching
 -------------------------------------------------------------------------------
 
-{-
+
 -- | Perform a binary search in the array to find an element.
-bsearch :: a -> Array a -> Maybe Bool
-bsearch = undefined
+{-# INLINE bsearch #-}
+bsearch :: (MonadIO m, Ord a , Storable a) => a -> Array a -> m Bool
+bsearch ele arr = go 0 (length arr - 1)
+  where
+    go low high
+        | high < low = return False
+        | otherwise = do
+            r <- readIndex arr mid
```

Comment:
`readIndex` is not an efficient way of reading because it performs bounds checks every time. See my comments in `readSlice` for the same.